src/openssl/mod.rs

```diff
@@ -310,20 +325,358 @@ impl ECVRF {
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * s, An element of field F such that s^2 == x
+    fn sqrt(&mut self, x: &BigNum) -> Result<BigNum, Error> {
+        let one = BigNum::from_u32(1)?;
+        let four = BigNum::from_u32(4)?;
+
+        let mut inv4 = BigNum::new()?;
+        inv4.mod_inverse(&four, &self.p, &mut self.bn_ctx)?;
+
+        let mut qp1 = BigNum::new()?;
+        qp1.mod_add(&self.p, &one, &self.p, &mut self.bn_ctx)?;
+        let mut qp1d4 = BigNum::new()?;
+        qp1d4.mod_mul(&qp1, &inv4, &self.p, &mut self.bn_ctx)?;
+
+        let mut result = BigNum::new()?;
+        result.mod_exp(x, &qp1d4, &self.p, &mut self.bn_ctx)?;
+        Ok(result)
+    }
+
+    /// Function to get the modular inverse of a field element in the field F
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * multiplicative inverse of x in F
+    fn inv0(&mut self, x: &BigNum) -> Result<BigNum, Error> {
+        let two = BigNum::from_u32(2)?;
+
+        let mut qm2 = BigNum::new()?;
+        qm2.mod_sub(&self.p, &two, &self.p, &mut self.bn_ctx)?;
+
+        let mut result = BigNum::new()?;
+        result.mod_exp(x, &qm2, &self.p, &mut self.bn_ctx)?;
+        Ok(result)
+    }
+
+    /// Function to get the sign of a field element
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * -1 or 1
+    fn sgn0(&mut self, x: &BigNum) -> Result<i32, Error> {
+        // This function is not yet generalized for (m > 1)
+        let mut sign = 0i32;
+        let zero = BigNum::from_u32(0)?;
+        let one = BigNum::from_u32(1)?;
+        let two = BigNum::from_u32(2)?;
+
+        let mut pm1 = BigNum::new()?;
+        pm1.mod_sub(&self.p, &one, &self.p, &mut self.bn_ctx)?;
+        let mut inv2 = BigNum::new()?;
+        inv2.mod_inverse(&two, &self.p, &mut self.bn_ctx)?;
+        let mut pm1d2 = BigNum::new()?;
+        pm1d2.mod_mul(&pm1, &inv2, &self.p, &mut self.bn_ctx)?;
+
+        let criterion = x.ucmp(&pm1d2) == Ordering::Greater;
+        let mut sign_0 = ECVRF::icmov(1, -1, criterion);
+        sign_0 = ECVRF::icmov(sign_0, 0, x.ucmp(&zero) == Ordering::Equal);
+        sign = ECVRF::icmov(sign, sign_0, sign == 0);
+
+        Ok(ECVRF::icmov(sign, 1, sign == 0))
+    }
+
+    /// Function to get the value of squared y on the curve
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * the value of x^3 + a * x + b
+    fn function_g(&mut self, x: &BigNum) -> Result<BigNum, Error> {
```

Comment:
This is not a intuitive function name. There could be a better name, like `get_squared_y`.

Reply:
OK. When a developer work only with this code, the `get_squared_y` name seems more intuitive. The original intention of the naming is to fit into the `irtf-cfrg` documentation.

Reply:
It is fine in the draft because there are terminologies, but not in here. That's the point.

Reply:
I renamed the function.